bin/run.sh

```diff
@@ -39,36 +35,51 @@ touch filter.conf
 export https_proxy=127.0.0.1:4343
 tinyproxy -d -c proxy.conf &
 
-# Un-skip all the skipped tests
-sed -i 's/skip <|//g' tests/Tests.elm
-
 # Temporarily disable -e mode
 set +e
-elm-test-rs -v --report exercism --offline > $OUTPUT_DIR/results.json 2> stderr.txt
+# Un-skip all the skipped tests
+sed -i 's/skip <|//g' tests/Tests.elm 2> stderr.txt
+STATUS=$?
+cat stderr.txt
+if [ $STATUS -ne 0 ]; then
+    jq -n --rawfile m stderr.txt '{version: 3, status: "error", message:$m}' > $OUTPUT_DIR/results.json
+    echo "An error occured while un-skipping the tests."
+    exit 0
+fi
+
+# Run the tests
+elm-test-rs -v --report exercism --offline > test_results.json 2> stderr.txt
 STATUS=$?
 cat stderr.txt
 # elm-test-rs will exit(0) if tests pass, exit(2) if tests fail
 if [ $STATUS -ne 0 ] && [ $STATUS -ne 2 ]; then
-   jq -n --rawfile m stderr.txt '{version: 3, status: "error", message:$m}' > $OUTPUT_DIR/results.json
-   echo "Finished with error"
-   exit 0
+    jq -n --rawfile m stderr.txt '{version: 3, status: "error", message:$m}' > $OUTPUT_DIR/results.json
+    echo "An error occured while running the tests."
+    exit 0
 fi
 
 # Extract test code
-cat tests/Tests.elm | node ../bin/cli.js > $OUTPUT_DIR/test_code.json 2> stderr.txt
+cat tests/Tests.elm | node /opt/test-runner/bin/cli.js > test_code.json 2> stderr.txt
 STATUS=$?
 cat stderr.txt
 if [ $STATUS -ne 0 ]; then
-   echo "An error occurred while extracting the test code from the test file."
-   exit 0
+    jq -n --rawfile m stderr.txt '{version: 3, status: "error", message:$m}' > $OUTPUT_DIR/results.json
+    echo "An error occurred while extracting the test code snippets."
+    exit 0
+fi
+
+# Check number of tests matches number of extracted code snippets
+test_code_length=$(jq 'length' test_code.json)
+test_result_length=$(jq '.tests | length' test_results.json)
+if [ $test_code_length -ne $test_result_length ] ; then
+    err="Number of tests doesn't match number of extracted code snippets. Please report this issue at https://github.com/exercism/elm-test-runner/issues."
+    jq -n --arg m "${err}" '{version: 3, status: "error", message:$m}' > $OUTPUT_DIR/results.json
+    echo $err
+    exit 0
 fi
 set -e
 
-cd $OUTPUT_DIR
-# Merge tests results with extracted test code.
-# This rely on the fact that the order is the same in both arrays.
-jq -s '[range(.[1]|length) as $i | .[0].tests[$i] + { test_code: .[1][$i].testCode }]' results.json test_code.json > concat.json
-jq -s '.[0].tests = .[1] | .[0]' results.json concat.json > results_with_code.json
-mv results_with_code.json results.json
+# Merge tests results with extracted test code, in the extracted code order.
+jq --slurpfile code test_code.json '.tests = (.tests + ($code | .[0] | to_entries | map({order: .key, name: .value.name, test_code: .value.testCode})) | group_by(.name) | map(add) | sort_by(.order) | map(del(.order)))' test_results.json > $OUTPUT_DIR/results.json
```

Comment:
Thanks for the jq magic!

Reply:
It was quite fun to put together, I had barely used jq before, now I'm seeing a glimpse of its power.